### PR TITLE
watch should never rebuild on changes that are ignored

### PR DESCRIPTION
**Pull request by nott**
There was a bug when `path` contained `./` as prefix (e.g. `./build/index.html`). Now `path::PathBuf` takes care of corner cases like this.

The bug described above meant there was an infinite rebuild loop in `cobalt watch` even when `build` directory was marked as ignored:

* on each rebuild some files in `build` directory were updated
* `inotify` started new rebuilds on these updates

### CODE REVIEW

**Comment by nott** [comment]
I don't have ideas for proper unit tests here. Do we need to add some?

**Comment by johannhof** [comment]
Hey there! Thanks for fixing this! I'll take a look later today. Testing sounds tricky, even with our new [CLI tests](https://github.com/cobalt-org/cobalt.rs/blob/master/tests/cli.rs). So I don't mind missing tests here :)

File: src/main.rs [hunk 1]
```diff
@@ -287,28 +287,29 @@ fn main() {
...
-                                                Pattern::matches_path(pattern, rel_path)
-                                            });
-                                        if !path_starts_with_build {
-                                            build(&config);
+                                    // get where process was run from
+                                    let cwd = std::env::current_dir().unwrap_or(PathBuf::new());
+
+                                    // the final goal is to have a relative path
+                                    // without corner-case prefixes like "./"
+                                    let abs_path = match path.is_absolute() {
```
Comment (johannhof) [inline src/main.rs]: You can probably replace this with

```rust
let abs_path = if path.is_absolute() {
  ...
} else {
  ...
}
```

File: src/main.rs [hunk 2]
```diff
@@ -287,28 +287,29 @@ fn main() {
...
+                                        true => path.clone(),
```
Comment (johannhof) [inline src/main.rs]: Why do you have to clone here?

File: src/main.rs [hunk 3]
```diff
@@ -287,28 +287,29 @@ fn main() {
...
+                                        false => {
+                                            // let path::Path handle corner cases like `./rel/path`
+                                            let mut abs_path = cwd.clone();
+                                            abs_path.push(path.clone());
+                                            abs_path
```
Comment (johannhof) [inline src/main.rs]: I think this whole block can be cut to `cwd.join(path)`

File: src/main.rs [hunk 4]
```diff
@@ -287,28 +287,29 @@ fn main() {
...
-                                            .any(|pattern| Pattern::matches_path(pattern, &path)) {
-                                            build(&config);
-                                        }
+                                    // check whether this path has been marked as ignored in config
+                                    let rel_path_matches =
+                                        |pattern| Pattern::matches_path(pattern, rel_path);
+                                    let path_ignored = &config.ignore.iter().any(rel_path_matches);
+
+                                    if !path_ignored {
+                                        build(&config);
```
Comment (johannhof) [inline src/main.rs]: Thanks, this section is much clearer now.

File: src/main.rs (hunk 3 above)
Comment (nott) [inline src/main.rs]: nice idea

File: src/main.rs (hunk 2 above)
Comment (nott) [inline src/main.rs]: `&path` is used in `unwrap_or` of `rel_path`, so I have to `path.clone()` here to have same datatype to `true` and `false branches.

**Comment by nott** [comment]
@johannhof thank you for the feedback. Everything but one `path.clone()` has been addressed in 86307d6.

**Comment by johannhof** [comment]
Looks good, thanks! We should probably also add the destination directory in there by default, to avoid infinite loops in case the destination is in the source folder, but that's for another pull request.

Merged 👍 ❤️

**Pull request merged**